examples/recording_service/leveldb-integration/cpp/LevelDbWriter.cxx

```diff
@@ -0,0 +1,512 @@
+            && ex != DDS_NO_EXCEPTION_CODE) {
+        value_buffer_.resize(8192);
+    } else {
+        DDS_UnsignedLong size = DDS_TypeCode_get_cdr_serialized_sample_max_size(
+                type,
+                DDS_AUTO_DATA_REPRESENTATION,
+                &ex);
+        if (ex != DDS_NO_EXCEPTION_CODE) {
+            std::stringstream log_stream;
+            log_stream << "Could not get serialized sample max size for stream"
+                    << std::endl;
+            log_stream << "    Stream name = " << stream_info.stream_name()
+                    << std::endl;
+            throw std::runtime_error(log_stream.str());
+        }
+        value_buffer_.resize((std::vector<char>::size_type) size);
+    }
+}
+
+LevelDbStreamWriter::~LevelDbStreamWriter()
+{
+}
+
+void LevelDbStreamWriter::store(
+        const std::vector<dds::core::xtypes::DynamicData *>& sample_seq,
+        const std::vector<dds::sub::SampleInfo *>& info_seq)
+{
+    using namespace dds::core::xtypes;
+    using namespace rti::core::xtypes;
+    using namespace dds::sub;
+
+    leveldb::WriteBatch transaction;
+
+    const int32_t count = sample_seq.size();
+    for (int32_t i = 0; i < count; ++i) {
+        /*
+         * The key of a sample has to be unique, and is composed of the
+         * following:
+         * - the reception timestamp
+         * - the original writer virtual GUID
+         * - the original virtual sequence number
+         */
+        const SampleInfo& sample_info = *(info_seq[i]);
+        int64_t reception_timestamp = sample_info->reception_timestamp().sec();
+        reception_timestamp *= rti::core::nanosec_per_sec;
+        reception_timestamp += sample_info->reception_timestamp().nanosec();
+        dds::core::array<uint8_t, GUID_LENGTH> guid;
+        rti::core::Guid original_pub_v_guid(
+                sample_info.delegate().original_publication_virtual_guid());
+        memcpy(guid.data(), original_pub_v_guid.native().value, GUID_LENGTH);
+        rti::core::SequenceNumber seq_nr(
+                sample_info.delegate().original_publication_virtual_sequence_number());
+        UserDataKey key(
+                SequenceNumber(seq_nr.high(), seq_nr.low()),
+                reception_timestamp,
+                guid);
+        dds::topic::topic_type_support<UserDataKey>::to_cdr_buffer(
+                key_buffer_,
+                key);
+        leveldb::Slice key_slice(key_buffer_.data(), key_buffer_.size());
+        /*
+         * The user-data sample is stored along with valid data flag. In fact,
+         * if the sample is not valid we do not serialize the sample to a CDR
+         * buffer at all
+         */
+        UserDataValue value;
+        value.valid_data(sample_info.valid());
+        if (sample_info.valid()) {
+            const dds::core::xtypes::DynamicData& sample = *(sample_seq[i]);
+            rti::core::xtypes::to_cdr_buffer(value.data_blob(), sample);
```

Comment:
this operation will resize data_blob() as needed, which is fine for the example. Then, why bother with the unbounded situation for value_buffer_?

Reply:
In the bounded case, we will avoid memory allocations in the critical path, by setting the buffer's capacity to the max serialized sample's size.